src/hotspot/share/opto/superword.cpp

```diff
@@ -411,6 +416,139 @@ void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
+  // Test whether there is a reduction cycle via every edge index
+  // (typically indices 1 and 2).
+  for (uint input = 1; input < n->req(); input++) {
+    if (in_reduction_cycle(n, input)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool SuperWord::is_reduction_operator(const Node* n) {
+  int opc = n->Opcode();
+  return (opc != ReductionNode::opcode(opc, n->bottom_type()->basic_type()));
+}
+
+bool SuperWord::in_reduction_cycle(const Node* n, uint input) {
+  // First find input reduction path to phi node.
+  auto has_my_opcode = [&](const Node* m){ return m->Opcode() == n->Opcode(); };
+  PathEnd path_to_phi = find_in_path(n, input, LoopMaxUnroll, has_my_opcode,
+                                     [&](const Node* m) { return m->is_Phi(); });
+  const Node* phi = path_to_phi.first;
+  if (phi == nullptr) {
+    return false;
+  }
+  // If there is an input reduction path from the phi's loop-back to n, then n
+  // is part of a reduction cycle.
+  const Node* first = phi->in(LoopNode::LoopBackControl);
+  PathEnd path_from_phi = find_in_path(first, input, LoopMaxUnroll, has_my_opcode,
+                                       [&](const Node* m) { return m == n; });
+  return path_from_phi.first != nullptr;
+}
+
+Node* SuperWord::original_input(const Node* n, uint i) {
+  if (n->has_swapped_edges()) {
+    assert(n->is_Add() || n->is_Mul(), "n should be commutative");
+    if (i == 1) {
+      return n->in(2);
+    } else if (i == 2) {
+      return n->in(1);
+    }
+  }
+  return n->in(i);
+}
+
+void SuperWord::mark_reductions() {
+
+  _loop_reductions.clear();
+
+  // Iterate through all phi nodes associated to the loop and search for
+  // reduction cycles in the basic block.
+  for (DUIterator_Fast imax, i = lp()->fast_outs(imax); i < imax; i++) {
+    const Node* phi = lp()->fast_out(i);
+    if (!phi->is_Phi()) {
+      continue;
+    }
+    if (phi->outcnt() == 0) {
+      continue;
+    }
+    if (phi == iv()) {
+      continue;
+    }
+    // The phi's loop-back is considered the first node in the reduction cycle.
+    const Node* first = phi->in(LoopNode::LoopBackControl);
+    if (first == nullptr) {
+      continue;
+    }
+    // Test that the node fits the standard pattern for a reduction operator.
+    if (!is_reduction_operator(first)) {
+      continue;
+    }
+    // Test that 'first' is the beginning of a reduction cycle ending in 'phi'.
+    // To contain the number of searched paths, assume that all nodes in a
+    // reduction cycle are connected via the same edge index, modulo swapped
+    // inputs. This assumption is realistic because reduction cycles usually
+    // consist of nodes cloned by loop unrolling.
+    int reduction_input = -1;
+    int path_nodes = -1;
+    for (uint input = 1; input < first->req(); input++) {
+      // Test whether there is a reduction path in the basic block from 'first'
+      // to the phi node following edge index 'input'.
+      PathEnd path =
+        find_in_path(
```

Comment:
One way to add fault-tolerance to find_in_path could be to follow strict DFS semantics where an alternate path is taken if node's predicates are not satisfied, currently we are starting all over again from the first node of chain with a different reduction_input which prevents inferring reduction chain even though all the nodes in the chain are commutative isomorphic operations.

Reply:
Thanks for the observation and the suggestion! This changeset proposes swapped edge tracking for efficiency and better worst-case behavior of the analysis, but as you observe this is done at the expense of robustness. I will investigate whether the swapped edge tracking approach can be extended to deal with the `MinI`/`MaxI` transformations that you mention. If not, I will re-consider using a generic search approach like you and @eme64 suggest.

Reply:
Hi again @jatin-bhateja,

I have studied now [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673) in more detail, and my conclusion is that it is not a duplicate but actually orthogonal to this changeset. Even a perfect reduction analysis alone would not re-enable the missing vectorization, because the canonicalization transformations done by `MaxI/MinINode::Ideal()` inhibit SuperWord analysis at a later stage.

In light of this, I propose to re-open [JDK-8302673](https://bugs.openjdk.org/browse/JDK-8302673), and address it by handling all four combinations of two-level inputs in `MaxI/MinINode::Ideal()` instead of canonicalizing `MaxI/MinI` chains. This solution is, in my opinion, more straightforward (and not necessarily more expensive). The main reason is that it separates concerns, making it possible to reason about input swapping, optimization of `MaxI/MinI` nodes, reduction analysis, and auto-vectorization separately. I have [a WIP, prototype implementation](https://github.com/openjdk/jdk/compare/master...robcasloz:jdk:JDK-8302673) which seems to work fine for all of the discussed reduction analysis strategies. @jatin-bhateja, if you want I can take over JDK-8302673 and submit it for review once I have polished it.

Given that JDK-8302673 is orthogonal to this RFE, and that a solution to JDK-8302673 is available for which this RFE detects MaxI/MinI reductions correctly, I suggest to move on with this RFE as-is, and file a follow-up RFE to investigate generic search approaches.